packages/indexer-common/src/subgraphs.ts

```diff
@@ -323,3 +327,162 @@ export function isDeploymentWorthAllocatingTowards(
+  error(msg: string, o?: object, ...args: any[]): void
+  warn(msg: string, o?: object, ...args: any[]): void
+}
+
+export interface SubgraphQueryInterface {
+  query<Data = any>(
+    query: DocumentNode,
+    variables?: Record<string, any>,
+  ): Promise<QueryResult<Data>>
+}
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+const blockNumberQuery = gql`
+  {
+    _meta {
+      block {
+        number
+      }
+    }
+  }
+`
+
+interface BlockNumberInterface {
+  data: { _meta: { block: { number: number } } }
+  errors?: any
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Variables = Record<string, any>
+
+export class SubgraphFreshnessChecker {
+  subgraphName: string
+  provider: ProviderInterface
+  threshold: number
+  logger: LoggerInterface
+  sleepDurationMillis: number
+  retries: number
+
+  constructor(
+    subgraphName: string,
+    provider: ProviderInterface,
+    freshnessThreshold: number,
+    sleepDurationMillis: number,
+    logger: LoggerInterface,
+    retries: number,
+  ) {
+    this.subgraphName = subgraphName
+    this.provider = provider
+    this.threshold = freshnessThreshold
+    this.sleepDurationMillis = sleepDurationMillis
+    this.logger = logger
+    this.retries = retries
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public async checkedQuery<Data = any>(
+    subgraph: SubgraphQueryInterface,
+    query: DocumentNode,
+    variables?: Variables,
+  ): Promise<QueryResult<Data>> {
+    // Try to inject the latest block number into the original query.
+    let updatedQuery = query
+    try {
+      updatedQuery = mergeSelectionSets(query, blockNumberQuery)
+    } catch (err) {
+      const errorMsg = `Failed to append block number into ${this.subgraphName} query`
+      this.logger.error(errorMsg, { subgraph: this.subgraphName, query: print(query) })
+      throw new Error(errorMsg)
+    }
+
+    // Try obtaining a fresh subgraph query at most 10 times
+    return this.checkedQueryRecursive<Data>(
+      updatedQuery,
+      subgraph,
+      this.retries,
+      variables,
+    )
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private async checkedQueryRecursive<Data = any>(
+    updatedQuery: DocumentNode,
+    subgraph: SubgraphQueryInterface,
+    retriesLeft: number,
+    variables?: Variables,
+  ): Promise<QueryResult<Data>> {
+    if (retriesLeft < 0) {
+      const errorMsg = `Max retries reached for ${this.subgraphName} freshness check`
+      this.logger.error(errorMsg, {
+        subgraph: this.subgraphName,
+        query: print(updatedQuery),
+      })
+      throw new Error(errorMsg)
+    }
+
+    // Obtain the latest network block number and subgraph query in parallel.
+    const subgraphQueryPromise = subgraph.query(updatedQuery, variables) as Promise<
+      QueryResult<Data> & BlockNumberInterface
+    >
+    const latestNetworkBLockPromise = this.provider.getBlockNumber()
```

Comment:
`latestNetworkBLockPromise` cases to `latestNetworkBlockPromise`?

Reply:
I'm not sure I understand the question.

Reply:
just a subtle difference in casing: `BLock` to `Block`. I don't know if there's a convention to use `L` instead of `l` in `block`